### PR TITLE
🎉 New Source: Coin API

### PR DESCRIPTION
**Pull request by isaacharrisholt**
## What

It's often useful for FinTech companies to have access to market data. This connector adds data from Coin API to allow this.

## How

This connector adds the following streams:

- `ohlcv_historical_data`
- `trades_historical_data`

## Recommended reading order

1. `coin_api.yaml`
2. `spec.yaml`

## 🚨 User Impact 🚨

No changes to existing code.

## Pre-merge Checklist

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

None.

</details>

<details><summary><strong>Integration</strong></summary>

image

</details>

<details><summary><strong>Acceptance</strong></summary>

image

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-coin-api/integration_tests/catalog.json [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  "streams": [
+    {
+      "name": "ohlcv_historical_data",
+      "supported_sync_modes": ["full_refresh"],
+      "source_defined_cursor": true,
+      "default_cursor_field": "time_start",
+      "json_schema": {
+        "$schema": "http://json-schema.org/draft-07/schema#",
+        "type": "object",
+        "properties": {
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-coin-api/integration_tests/catalog.json:11]: you can remove this file.

File: airbyte-integrations/connectors/source-coin-api/integration_tests/configured_catalog.json [hunk 2]
```diff
@@ -0,0 +1,13 @@
+{
+  "streams": [
+    {
+      "stream": {
+        "name": "ohlcv_historical_data",
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-coin-api/integration_tests/configured_catalog.json:5]: can you add the second stream too and run the tests again?

File: airbyte-integrations/connectors/source-coin-api/source_coin_api/coin_api.yaml [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    http_method: "GET"
+    authenticator:
+      type: ApiKeyAuthenticator
+      header: "X-CoinAPI-Key"
+      api_token: "{{ config['api_key'] }}"
+    request_options_provider:
+      request_parameters:
+        period_id: "{{ config['period'] }}"
+        time_start: "{{ config['start_time'] }}"
+        time_end: "{{ config['end_time'] }}"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-coin-api/source_coin_api/coin_api.yaml:18]:
```suggestion
        time_end: "{{ config.get("end_time", now_utc()) }}"
```
not completly sure if works if not `config.get("end_time") or now_utc()`

File: airbyte-integrations/connectors/source-coin-api/source_coin_api/coin_api.yaml [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        limit: "{{ config['limit'] }}"
+  stream_slicer:
+    type: DatetimeStreamSlicer
+    start_datetime:
+      datetime: "{{ config['start_time'] }}"
+      datetime_format: "%Y-%m-%dT%H:%M:%S"
+    end_datetime:
+      datetime: "{{ config['end_time'] }}"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-coin-api/source_coin_api/coin_api.yaml:26]: same case here.

File: airbyte-integrations/connectors/source-coin-api/source_coin_api/spec.yaml [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      examples:
+        - "2019-01-01T00:00:00"
+    limit:
+      type: integer
+      description: |
+        The maximum number of elements to return. If not supplied, the default
+        is 100. For numbers larger than 100, each 100 items is counted as one
+        request for pricing purposes. Maximum value is 100000.
+      minimum: 1
+      maximum: 100000
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-coin-api/source_coin_api/spec.yaml:59]: add default.

File: airbyte-integrations/connectors/source-coin-api/source_coin_api/spec.yaml [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        - "2019-01-01T00:00:00"
+    end_time:
+      type: string
+      pattern: "^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}$"
+      description: |
+        The end time in ISO 8601 format. If not supplied, data will be returned
+        from the start time to the current time, or when the count of result
+        elements reaches its limit.
+      examples:
+        - "2019-01-01T00:00:00"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-coin-api/source_coin_api/spec.yaml:51]: start_date and end_date

File: airbyte-integrations/connectors/source-coin-api/source_coin_api/spec.yaml [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+          - production
+      default: sandbox
+      order: 1
+    symbol_id:
+      type: string
+      description: The symbol ID to use. See the documentation for a list.
+      order: 2
+    period:
+      type: string
+      description: The period to use. See the documentation for a list.
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-coin-api/source_coin_api/spec.yaml:33]: Can you add the link to the doc?

File: airbyte-integrations/connectors/source-coin-api/source_coin_api/spec.yaml [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+      description: |
+        The environment to use. Either sandbox or production.
+      enum:
+          - sandbox
+          - production
+      default: sandbox
+      order: 1
+    symbol_id:
+      type: string
+      description: The symbol ID to use. See the documentation for a list.
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-coin-api/source_coin_api/spec.yaml:29]: Link to the documentation.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
some comments, please add the documentatino for this connector. Also can you link to the coinbase documentation where you're retrieving this data?

Do you have a sandbox account to help us testing this connector?

File: airbyte-integrations/connectors/source-coin-api/source_coin_api/spec.yaml (hunk 8 above)
Comment (isaacharrisholt) [reply airbyte-integrations/connectors/source-coin-api/source_coin_api/spec.yaml]: Done

File: airbyte-integrations/connectors/source-coin-api/source_coin_api/spec.yaml (hunk 6 above)
Comment (isaacharrisholt) [reply airbyte-integrations/connectors/source-coin-api/source_coin_api/spec.yaml]: Done

File: airbyte-integrations/connectors/source-coin-api/source_coin_api/spec.yaml (hunk 5 above)
Comment (isaacharrisholt) [reply airbyte-integrations/connectors/source-coin-api/source_coin_api/spec.yaml:63]: Done

File: airbyte-integrations/connectors/source-coin-api/source_coin_api/coin_api.yaml (hunk 3 above)
Comment (isaacharrisholt) [reply airbyte-integrations/connectors/source-coin-api/source_coin_api/coin_api.yaml]: It works even if `end_date` is not provided.

File: airbyte-integrations/connectors/source-coin-api/source_coin_api/coin_api.yaml (hunk 4 above)
Comment (isaacharrisholt) [reply airbyte-integrations/connectors/source-coin-api/source_coin_api/coin_api.yaml:26]: As above

File: airbyte-integrations/connectors/source-coin-api/integration_tests/catalog.json (hunk 1 above)
Comment (isaacharrisholt) [reply airbyte-integrations/connectors/source-coin-api/integration_tests/catalog.json]: Done

File: airbyte-integrations/connectors/source-coin-api/integration_tests/configured_catalog.json (hunk 2 above)
Comment (isaacharrisholt) [reply airbyte-integrations/connectors/source-coin-api/integration_tests/configured_catalog.json:5]: Yep, oversight on my part, sorry.

image

**Comment by isaacharrisholt** [comment]
CoinAPI has a free plan that should allow you to test the connector.

**Review by marcosmarxm — APPROVED** [review]
Congrats @isaacharrisholt I'll publish the connector later.

**Comment by marcosmarxm** [comment]
Manually generated the source spec.

**Pull request merged**